### ISSUE TITLE
Cannot perform classification on custom dataset

### ISSUE THREAD

**Issue by nikogamulin** [nicolas-chaulet/torch-points3d#486]
Hi,

I tried to create a custom dataset of superquadrics and adapt the code from ObjectClassificationRSConv.ipynb notebook in order to perform classification.

The SuperQuadricsClassification and SuperQuadricsClassificationDataset are defined as follows:

```
class SuperQuadricsClassification(Dataset):
    def __init__(self, dataset_size, points_count, dimension_max, do_normalize=True, transform=None,
                 classification=True):
        """
        Args:
            number of superquadrics to be gnerated.
        """
        super().__init__()

        used_combinations = []
        self.pos = np.zeros((dataset_size, points_count, 3))
        self.norm = np.zeros((dataset_size, points_count, 3))
        if classification:
            self.y = np.zeros((dataset_size, 1))
        else:
            self.y = np.zeros((dataset_size, 8))
        pbar = tqdm(total=dataset_size)
        idx = 0
        while idx < dataset_size:
            combination_is_valid = False
            while not combination_is_valid:
                a1, a2, a3, e1, e2, x0, y0, z0 = get_random_superquadric_parameters()
                current_parameters_str = '{}#{}#{}#{}#{}#{}#{}#{}'.format(a1, a2, a3, e1, e2, x0, y0, z0)
                if current_parameters_str not in used_combinations:
                    combination_is_valid = True
            used_combinations.append(current_parameters_str)
            superquadric = Superquadric(a1, a2, a3, e1, e2, x0, y0, z0, dimension_max)
            try:
                points, normals = superquadric.get_grid(points_count)
                if do_normalize:
                    max_offset = sq_dimensions_range[1] + sq_offset_range[1]
                    points = list(map(lambda x: [item / max_offset for item in x], points))

            except:
                # logger.warning(
                #     'Unable to generate superquadric with following parameters: a1 = {}, a2 = {}, a3 = {}, e1 = {}, e2 = {}, x0 = {}, y0 = {}, z0 = {}'.format(
                #         a1, a2, a3, e1, e2, x0, y0, z0))
                continue
            self.pos[idx] = np.array(points)
            self.norm[idx] = normals
            if classification:
                if e1 < 0.5 and e2 < 0.5:
                    self.y[idx] = 0.0
                elif e1 < 0.5 and e2 > 0.5:
                    self.y[idx] = 1.0
                elif e1 > 0.5 and e2 < 0.5:
                    self.y[idx] = 2.0
                else:
                    self.y[idx] = 3.0
            else:
                self.y[idx] = np.array([a1, a2, a3, e1, e2, x0, y0, z0])
            idx += 1
            pbar.update(1)
        pbar.close()

    def __len__(self):
        return len(self.pos)

    def __getitem__(self, idx):
        if torch.is_tensor(idx):
            idx = idx.tolist()

        norm = torch.from_numpy(self.norm[idx]).float()
        pos = torch.from_numpy(self.pos[idx]).float()
        y = torch.from_numpy(self.y[idx]).float()
        return Data(normal=norm, pos=pos, y=y)


class SuperQuadricsClassificationDataset(BaseDataset):
    def __init__(self, dataset_opt):
        super().__init__(dataset_opt)

        # number = dataset_opt.number
        # if str(number) not in self.AVAILABLE_NUMBERS:
        #     raise Exception("Only ModelNet10 and ModelNet40 are available")
        self.train_dataset = SuperQuadricsClassification(dataset_size=1, points_count=2048, dimension_max=305, transform=self.train_transform)
        self.test_dataset = SuperQuadricsClassification(dataset_size=1, points_count=2048, dimension_max=305, transform=self.test_transform)

    def get_tracker(self, wandb_log: bool, tensorboard_log: bool):
        """Factory method for the tracker
        Arguments:
            wandb_log - Log using weight and biases
            tensorboard_log - Log using tensorboard
        Returns:
            [BaseTracker] -- tracker
        """
        return ClassificationTracker(self, wandb_log=wandb_log, use_tensorboard=tensorboard_log)
```

the functions for point cloud generation are defined as follows:

```
def phi(x, y, t, a1=1, a2=1, alpha=1e-4):
    u = max(x/a1, y/a2)**2
    v = min(x/a1, y/a2)**2
    if t < alpha:
        return u * (((1 + (v/u)**(1/alpha))**alpha - 1) * t / alpha + 1)
    else:
        return u * (1 + (v/u)**(1/t))**t


def p(theta, a1, a2, eps):
    phi_tmp = phi(np.abs(np.cos(theta)), np.abs(np.sin(theta)), eps, a1, a2)
    return np.array([np.cos(theta), np.sin(theta)]) / np.sqrt(phi_tmp)


def get_split_angle(ax, ay, eps, alpha, beta):
    return get_split_angle_bis(ax, ay, eps, alpha, beta, alpha, beta)


def get_split_angle_bis(a1, a2, eps, alpha, beta, la, lb):
    theta = (la+lb)/2
    pb = p(beta, a1, a2, eps)
    pt = p(theta, a1, a2, eps)
    pa = p(alpha, a1, a2, eps)
    norm1 = np.linalg.norm(pt - pb)
    norm2 = np.linalg.norm(pa - pt)

    if abs(norm1 - norm2) < 1e-4:
        return theta
    elif norm1 > norm2:
        return get_split_angle_bis(a1, a2, eps, alpha, beta, theta, lb)
    else:
        return get_split_angle_bis(a1, a2, eps, alpha, beta, la, theta)


class SuperellipseSampler:

    def __init__(self, a1, a2, eps, delta, d_max):
        self.a1 = a1
        self.a2 = a2
        self.eps = eps
        self.d_max = d_max
        self.delta = delta

    def sample_se(self, alpha, beta):
        I = np.array([np.pi, np.arctan2(self.a2, -self.a1), np.pi/2, np.arctan2(self.a2, self.a1), 0,
                      np.arctan2(-self.a2, self.a1), -np.pi/2, np.arctan2(-self.a2, -self.a1), -np.pi])

        I = I[np.logical_and(I > alpha, I < beta)]
        I = np.concatenate(([beta], I, [alpha]))
        P = []
        for i in range(1, I.shape[0]):
            P.append(p(I[i-1], self.a1, self.a2, self.eps))
            P += self.sample_segment(I[i-1], I[i], 0)
        P.append(p(I[-1], self.a1, self.a2, self.eps))
        return P

    def sample_segment(self, alpha, beta, d):
        pb = p(beta, self.a1, self.a2, self.eps)
        pa = p(alpha, self.a1, self.a2, self.eps)
        norm = np.linalg.norm(pb - pa)
        if d > self.d_max or norm < self.delta:
            return []

        theta = get_split_angle(self.a1, self.a2, self.eps, alpha, beta)
        P = []
        P += self.sample_segment(alpha, theta, d+1)
        P.append(p(theta, self.a1, self.a2, self.eps))
        P += self.sample_segment(theta, beta, d+1)
        return P


def fexp(x, p):
    """a different kind of exponentiation"""
    return (np.sign(x) * (np.abs(x) ** p))


def tens_fld(A, B, C, P, Q, x0=0, y0=0, z0=0):
    """this module plots superquadratic surfaces with the given parameters"""
    phi, theta = np.mgrid[0:np.pi:80j, 0:2 * np.pi:80j]
    x = A * (fexp(np.sin(phi), P)) * (fexp(np.cos(theta), Q)) + x0
    y = B * (fexp(np.sin(phi), P)) * (fexp(np.sin(theta), Q)) + y0
    z = C * (fexp(np.cos(phi), P)) + z0
    return x, y, z

def sample_sq(ax, ay, az, eps1, eps2, delta, d_max, ignore_center=False):

    se_xy_samp = SuperellipseSampler(ax, ay, eps2, delta, d_max)
    se_z_samp = SuperellipseSampler(1, az, eps1, delta, d_max)

    se_xy = se_xy_samp.sample_se(-np.pi, np.pi)
    diff = np.tan(delta/az/4) if ignore_center else 0
    se_z = se_z_samp.sample_se(-np.pi/2+diff, np.pi/2-diff)

    se_xy = np.array(se_xy)
    se_z = np.array(se_z)

    # Spherical product
    X = np.outer(se_xy[:, 0], se_z[:, 0])
    Y = np.outer(se_xy[:, 1], se_z[:, 0])
    Z = np.outer(np.ones(se_xy.shape[0]), se_z[:, 1])
    return X, Y, Z

class Superquadric:
    def __init__(self, a1, a2, a3, e1, e2, x0=0, y0=0, z0=0, dimension_max=0):
        """
        :param a1: size x
        :param a2: size y
        :param a3: size z
        :param e1: shape param1
        :param e2: shape param2
        :param x0: offset x
        :param y0: offset y
        :param z0: offset z
        """
        self.a1 = a1
        self.a2 = a2
        self.a3 = a3
        self.e1 = e1
        self.e2 = e2
        self.x0 = x0
        self.y0 = y0
        self.z0 = z0
        self.dimension_max = dimension_max

    def get_z_partial_param(self, z):
        z_partial = (1 - (z/self.a3)**(2/self.e1))**(self.e2/self.e1)
        return z_partial

    def get_x_value(self, y, z):
        z_partial = self.get_z_partial_param(z)
        a = (z_partial - (y/self.a2)**(2/self.e2))**self.e2
        if a < 0:
            return None
        x = self.a1 * math.sqrt(a)
        return x

    def get_y_value(self, x, z):
        z_partial = self.get_z_partial_param(z)
        a = (z_partial - (x/self.a1)**(2/self.e2))
        a = float(a)**self.e2
        if isinstance(a, complex):
            return None
        if a < 0:
            return None
        y = self.a2 * math.sqrt(a)
        return y

    def get_z_value(self, x, y):
        a = ((x/self.a1) ** (2/self.e2) + (y/self.a2) ** (2/self.e2)) ** (self.e2/self.e1)
        if a > 1:
            return None
        z = self.a3 * math.sqrt((1 - a) ** self.e1)
        return z

    def get_grid(self, sample_size, include_normals=True, draw_grid=False):
        normals = None
        # x, y, z = tens_fld(self.a1, self.a2, self.a3, self.e1, self.e2, self.x0, self.y0, self.z0)
        x, y, z = sample_sq(ax=self.a1, ay=self.a2, az=self.a3, eps1=self.e1, eps2=self.e2, delta=0.5, d_max=10, ignore_center=False)
        x += self.x0
        y += self.y0
        z += self.z0
        x = x.reshape(-1)
        y = y.reshape(-1)
        z = z.reshape(-1)
        points = list(map(list, zip(x.flatten(), y.flatten(), z.flatten())))
        sampled_indices = random.sample(range(len(points)), sample_size)
        if include_normals:
            pcd = o3d.geometry.PointCloud()
            pcd.points = o3d.utility.Vector3dVector(points)
            pcd.estimate_normals()
            normals = list(np.asarray(pcd.normals))
        if draw_grid:
            o3d.visualization.draw_geometries([pcd], point_show_normal=True)
            # c = [a[i] for i in b]
        sampled_points = [points[i] for i in sampled_indices]
        sampled_normals = [normals[i] for i in sampled_indices]
        return sampled_points, sampled_normals

    @staticmethod
    def draw_point_cloud(xyz_tuples):
        fig = plt.figure(figsize=plt.figaspect(1))
        ax = fig.add_subplot(111, projection='3d')
        vals_x = [item[0] for item in xyz_tuples]
        vals_y = [item[1] for item in xyz_tuples]
        vals_z = [item[2] for item in xyz_tuples]
        ax.scatter(vals_x, vals_y, vals_z)
        plt.show()
        # plt.display()
        plt.waitforbuttonpress()
```

These are the examples of generated point clouds:

![sq_samples](https://user-images.githubusercontent.com/1803701/97590920-66facf00-19ff-11eb-8850-8d89ee7f0ad6.png)

The following is the test script to perform classification:

```
USE_NORMAL = True  # @param {type:"boolean"}
    DIR = ""
    yaml_config = """
    task: classification
    class: modelnet.ModelNetDataset
    name: modelnet
    dataroot: %s
    pre_transforms:
        - transform: NormalizeScale
        - transform: GridSampling3D
          lparams: [0.02]
    train_transforms:
        - transform: FixedPoints
          lparams: [2048]
        - transform: RandomNoise
        - transform: RandomRotate
          params:
            degrees: 180
            axis: 2
        - transform: AddFeatsByKeys
          params:
            feat_names: [norm]
            list_add_to_x: [%r]
            delete_feats: [True]
    test_transforms:
        - transform: FixedPoints
          lparams: [2048]
        - transform: AddFeatsByKeys
          params:
            feat_names: [norm]
            list_add_to_x: [%r]
            delete_feats: [True]
    """ % (os.path.join(DIR, "data"), USE_NORMAL, USE_NORMAL)

    from omegaconf import OmegaConf
    params = OmegaConf.create(yaml_config)
    dataset = SuperQuadricsClassificationDataset(params)
    # Setup the data loaders
    model = RSConvCLassifier()
    dataset.create_dataloaders(
        model,
        batch_size=BATCH_SIZE,
        shuffle=True,
        num_workers=NUM_WORKERS,
        precompute_multi_scale=False
    )
    print(next(iter(dataset.test_dataloaders[0])))
    # Setup the tracker and actiavte tensorboard loging
    logdir = "/home/niko/workspace/torch-points3d/runs"
    logdir = os.path.join(logdir, str(datetime.datetime.now()))
    os.mkdir(logdir)
    os.chdir(logdir)
    tracker = dataset.get_tracker(False, True)

    optimizer = torch.optim.Adam(model.parameters(), lr=0.01)
    EPOCHS = 50
    for i in range(EPOCHS):
        print("=========== EPOCH %i ===========" % i)
        time.sleep(0.5)
        train_epoch('cuda')
        tracker.publish(i)
        test_epoch('cuda')
        tracker.publish(i)



```

While running it, I get the following error:

```
SimpleBatch(normal=[1, 2048, 3], pos=[1, 2048, 3], y=[1, 1])
=========== EPOCH 0 ===========
  0%|          | 0/1 [00:00<?, ?it/s]SimpleBatch(normal=[1, 2048, 3], pos=[1, 2048, 3], y=[1, 1])
  0%|          | 0/1 [00:00<?, ?it/s]
Traceback (most recent call last):
  File "/home/niko/workspace/torch-points3d/torch_points3d/datasets/classification/superquadrics.py", line 46, in train_epoch
    model.forward(data)
  File "/home/niko/workspace/torch-points3d/torch_points3d/datasets/classification/superquadrics.py", line 109, in forward
    data_out = self.encoder(data)
  File "/home/niko/workspace/torch-points3d/venv/lib/python3.6/site-packages/torch/nn/modules/module.py", line 532, in __call__
    result = self.forward(*input, **kwargs)
  File "/home/niko/workspace/torch-points3d/torch_points3d/applications/rsconv.py", line 136, in forward
    data = self.down_modules[i](data)
  File "/home/niko/workspace/torch-points3d/venv/lib/python3.6/site-packages/torch/nn/modules/module.py", line 532, in __call__
    result = self.forward(*input, **kwargs)
  File "/home/niko/workspace/torch-points3d/torch_points3d/core/base_conv/dense.py", line 73, in forward
    ms_x.append(self.conv(x, pos, new_pos, radius_idx, scale_idx))
  File "/home/niko/workspace/torch-points3d/torch_points3d/modules/RSConv/dense.py", line 385, in conv
    aggr_features
  File "/home/niko/workspace/torch-points3d/venv/lib/python3.6/site-packages/torch/nn/modules/module.py", line 532, in __call__
    result = self.forward(*input, **kwargs)
  File "/home/niko/workspace/torch-points3d/torch_points3d/modules/RSConv/dense.py", line 250, in forward
    x = self.activation(self.bn_xyz_raising(self.xyz_raising(x)))
  File "/home/niko/workspace/torch-points3d/venv/lib/python3.6/site-packages/torch/nn/modules/module.py", line 532, in __call__
    result = self.forward(*input, **kwargs)
  File "/home/niko/workspace/torch-points3d/venv/lib/python3.6/site-packages/torch/nn/modules/conv.py", line 345, in forward
    return self.conv2d_forward(input, self.weight)
  File "/home/niko/workspace/torch-points3d/venv/lib/python3.6/site-packages/torch/nn/modules/conv.py", line 342, in conv2d_forward
    self.padding, self.dilation, self.groups)
RuntimeError: Given groups=1, weight of size 16 6 1 1, expected input[1, 3, 1024, 16] to have 6 channels, but got 3 channels instead

Process finished with exit code 1
```

I've been struggling with it for a long time so I would really appreciate it if anyone could suggest how to appropriately generate a dataset of point clouds in order to be suitable with implemented models or point out what might be the cause of the error pasted above.

Basically, I want to generate a dataset of [superquadrics](https://en.wikipedia.org/wiki/Superquadrics) and split them into multiple classes depending on their epsilon values (which determine the shape) and the try to correctly predict the class.

**Comment by nicolas-chaulet** [comment]
Hi @nikogamulin your dataset definitely looks correct but the config might be wrong. This line here
```
print(next(iter(dataset.test_dataloaders[0])))
```
should print a data object that has `pos`, `y` and `x` but instead of `x` you have `norm`. 
This bit in the dataset config:
```
        - transform: AddFeatsByKeys
          params:
            feat_names: [norm]
            list_add_to_x: [%r]
            delete_feats: [True]
```
should tell it to add `norm` to the feature vector `x` but it looks like it does not do it. Could you print the dataset itself (`print(dataset)`) and check that all is correct in it?

**Comment by nikogamulin** [comment]
hi @nicolas-chaulet , the database is the following:

Dataset: SuperQuadricsClassificationDataset 
pre_transform = Compose([
    NormalizeScale(),
    GridSampling3D(grid_size=0.02, quantize_coords=False, mode=mean),
])
test_transform = Compose([
    FixedPoints(2048, replace=True),
    AddFeatsByKeys(norm=True),
])
train_transform = Compose([
    FixedPoints(2048, replace=True),
    RandomNoise(sigma=0.01, clip=0.05),
    RandomRotate((-180, 180), axis=2),
    AddFeatsByKeys(norm=True),
])
val_transform = None
inference_transform = Compose([
    NormalizeScale(),
    GridSampling3D(grid_size=0.02, quantize_coords=False, mode=mean),
    FixedPoints(2048, replace=True),
    AddFeatsByKeys(norm=True),
])
Size of train_dataset = 1
Size of test_dataset = 1
Size of val_dataset = 0
Batch size = None

I have also uploaded a jupyter notebook with all the functions to generate the dataset and perform the training: https://github.com/nikogamulin/superquadrics/blob/main/custom-dataloader-v2.ipynb

Thank you for help!

**Comment by nicolas-chaulet** [comment]
I found the issue, the `data` object that you return stores the normal vector into the `normal` attribute (not `norm` as expected by the transform). So you need to either change your transform or your dataset. In any case, the data object returned by `dataset.train_dataset[0]` should contain a `x` attribute of size `[N,3]`. `x` is reserved for the features that get fed to the network.

**Comment by nikogamulin** [comment]
@nicolas-chaulet , thanks for help. I first renamed normal to norm but the train_dataset item didn't contain x - probably I overlooked something while trying to adapt the shape of my dataset to match ModelNetDataset. Then, however, I simply added x:

```
def __getitem__(self, idx):
        if torch.is_tensor(idx):
            idx = idx.tolist()
        norm = torch.from_numpy(self.norm[idx]).float()
        pos = torch.from_numpy(self.pos[idx]).float()
        y = torch.from_numpy(self.y[idx]).long()
        return Data(norm=norm, x=norm, pos=pos, y=y)
```



Next, I will try to make output value multidimensional as the end goal is to determine the body size, shape, and position. After 3 epochs, I get the following error:

```
Traceback (most recent call last):
  File "/home/niko/workspace/torch-points3d/venv/lib/python3.6/site-packages/torch/utils/data/dataloader.py", line 961, in __del__
    self._shutdown_workers()
  File "/home/niko/workspace/torch-points3d/venv/lib/python3.6/site-packages/torch/utils/data/dataloader.py", line 941, in _shutdown_workers
    w.join()
  File "/usr/lib/python3.6/multiprocessing/process.py", line 122, in join
    assert self._parent_pid == os.getpid(), 'can only join a child process'
AssertionError: can only join a child process
Exception ignored in: <bound method _MultiProcessingDataLoaderIter.__del__ of <torch.utils.data.dataloader._MultiProcessingDataLoaderIter object at 0x7f07e81ecc88>>
```

But the learning process seems to be running and the training and test accuracy increases.

Next, I am planning to generate a dataset with output values in the shape of 8-dimensional vectors (2 values parametrize the shape, 3 the size, and 3 the position in 3D space).

**Comment by nicolas-chaulet** [comment]
I believe this is fixed for now correct?

**Comment by ngamulin** [comment]
fixed.
